quinn-proto/src/connection/paths.rs

```diff
@@ -163,3 +177,605 @@ impl RttEstimator {
+                state.lost_probe_count = 0;
+                state.in_flight_probe = None;
+            }
+
+            if let Some(probe_udp_payload_size) = state.next_mtu_to_probe(last_probe_succeeded) {
+                state.in_flight_probe = Some(next_packet_number);
+                state.last_probed_udp_payload_size = probe_udp_payload_size;
+                return Some(probe_udp_payload_size);
+            } else {
+                let next_mtud_activation = now + self.config.interval;
+                self.phase = Phase::Complete(next_mtud_activation);
+                return None;
+            }
+        }
+
+        None
+    }
+
+    /// Called when a packet is acknowledged in [`SpaceId::Data`]
+    ///
+    /// Returns the new `current_mtu` if the packet number corresponds to the in-flight MTU probe
+    fn on_probe_acked(&mut self, packet_number: u64) -> Option<u16> {
+        match &mut self.phase {
+            Phase::Searching(state) if state.in_flight_probe == Some(packet_number) => {
+                state.in_flight_probe = None;
+                state.lost_probe_count = 0;
+                Some(state.last_probed_udp_payload_size)
+            }
+            _ => None,
+        }
+    }
+
+    /// Called when the in-flight MTU probe was lost
+    fn on_probe_packet_lost(&mut self) {
+        // We might no longer be searching, e.g. if a black hole was detected
+        if let Phase::Searching(state) = &mut self.phase {
+            state.in_flight_probe = None;
+            state.lost_probe_count += 1;
+        }
+    }
+
+    /// Called when a black hole is detected
+    fn on_black_hole_detected(&mut self, now: Instant) {
+        // Stop searching, if applicable, and reset the timer
+        let next_mtud_activation = now + self.config.black_hole_cooldown;
+        self.phase = Phase::Complete(next_mtud_activation);
+    }
+}
+
+#[derive(Debug, Clone)]
+enum Phase {
+    /// We haven't started polling yet
+    Initial,
+    /// We are currently searching for a higher PMTU
+    Searching(SearchState),
+    /// Searching has completed and will be triggered again at the provided instant
+    Complete(Instant),
+}
+
+#[derive(Debug, Clone)]
+struct SearchState {
+    /// The lower bound for the current binary search
+    search_lower_bound: u16,
+    /// The upper bound for the current binary search
+    search_upper_bound: u16,
+    /// The UDP payload size we last sent a probe for
+    last_probed_udp_payload_size: u16,
+    /// Packet number of an in-flight probe (if any)
+    in_flight_probe: Option<u64>,
+    /// Lost probes at the current probe size
+    lost_probe_count: usize,
+}
+
+impl SearchState {
+    /// Creates a new search state, with the specified lower bound (the upper bound is derived from
+    /// the config and the peer's `max_udp_payload_size` transport parameter)
+    fn new(lower_bound: u16, peer_max_udp_payload_size: u16, config: &PlpmtudConfig) -> Self {
+        let search_upper_bound = config
```

Comment:
Consider using `clamp()` here?

Reply:
Unfortunately, `clamp` will panic if `min > max`, which is something we should handle robustly here (i.e. we want `config.upper_bound.clamp(lower_bound, peer_max_udp_payload_size)`, but it is possible that `lower_bound` is higher than `peer_max_udp_payload_size`).

Reply:
Fair enough, maybe add a comment about this?

Reply:
While adding the comment I figured out a way to use clamp after all, just pushed it ;)